### PR TITLE
fix: Fix transcoding and release as 4.1.1

### PR DESCRIPTION
**Pull request by jskeet**
It's definitely worth reviewing one commit at a time.
Most commits are small in terms of production code, but the generated test code can be large - skip the generated code.

### CODE REVIEW

File: Google.Api.Gax.Grpc/Rest/RestServiceCollection.cs [hunk 1]
```diff
@@ -37,7 +37,8 @@ internal static RestServiceCollection Create(ApiMetadata metadata)
             var typeRegistry = TypeRegistry.FromFiles(fileDescriptors.ToArray());
             var parser = new JsonParser(JsonParser.Settings.Default.WithIgnoreUnknownFields(true).WithTypeRegistry(typeRegistry));
             var methodsByName = services.SelectMany(service => service.Methods)
-                .Where(x => true) // TODO: filter out streaming methods.
+                // We don't yet support streaming methods.
+                .Where(x => !x.IsClientStreaming && !x.IsServerStreaming)
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/RestServiceCollection.cs:41]: Both of these will be true for bidi I think, else we need a specific check.

File: Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs [hunk 2]
```diff
@@ -268,12 +270,44 @@ internal IEnumerable<string> GetValues(IMessage request)
...
 
             static string FormatValue(object value) => 
                 value is bool b ? (b ? "true" : "false")
+                : value is Enum enumValue ? OriginalEnumValueHelper.GetOriginalName(enumValue)
                 : value is IFormattable formattable ? formattable.ToString(format: null, CultureInfo.InvariantCulture)
                 : value.ToString();
         }
+
+        // Effectively a cache of mapping from enum values to the original name as specified in the proto file,
+        // fetched by reflection. This code is taken from the protobuf JsonFormatter.
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs]: I'm guessing the protobuf JsonFormatter is private then, or this code is at least?

File: Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs [hunk 3]
```diff
@@ -170,27 +170,19 @@ internal FieldSegment(string fieldText, MessageDescriptor descriptor)
                 descriptor = field.MessageType;
             }
 
-            // Now handle the last field, which must be a singular integer or string field.
+            // Now handle the last field, which must be a singular scalar field.
+            // (It's extremely unusual for it to be anything other than a string field, but it's easy enough to handle.)
             var lastFieldName = fieldNames.Last();
             var lastField = GetField(descriptor, lastFieldName);
+            if (lastField.IsRepeated || lastField.FieldType == FieldType.Message)
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs:177]: Is `bytes` allowed as well?

File: Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs [hunk 4]
```diff
@@ -268,46 +268,46 @@ internal IEnumerable<string> GetValues(IMessage request)
...
-        // Effectively a cache of mapping from enum values to the original name as specified in the proto file,
-        // fetched by reflection. This code is taken from the protobuf JsonFormatter.
-        private static class OriginalEnumValueHelper
-        {
-            private static readonly ConcurrentDictionary<System.Type, Dictionary<object, string>> _dictionaries
-                = new ConcurrentDictionary<System.Type, Dictionary<object, string>>();
+    internal static string FormatPathOrQueryValue(object value) =>
+        value is bool b ? (b ? "true" : "false")
+        : value is Enum enumValue ? OriginalEnumValueHelper.GetOriginalName(enumValue)
+        : value is IFormattable formattable ? formattable.ToString(format: null, CultureInfo.InvariantCulture)
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs:282]: For the floating point numbers, is everyone else using invariant culture, or server side at least? Is that one of the things that's unclear?

**Review by amanda-tarafa — APPROVED** [review]
LGTM

File: Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs (hunk 4 above)
Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs:282]: Everyone should definitely be using the invariant culture or equivalent (no "1,5" for example) but there are other problems around precision. Happy to talk about it on a VC if that would be useful.

File: Google.Api.Gax.Grpc/Rest/RestServiceCollection.cs (hunk 1 above)
Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/RestServiceCollection.cs:41]: Yup, I'm pretty sure they're both true for bidi.

File: Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs (hunk 2 above)
Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/HttpRuleTranscoder.cs]: This code is private, yes. It might be good to expose it at some point, but that can wait.

**Pull request merged**